addons/point_of_sale/models/stock_picking.py

```diff
@@ -83,81 +84,111 @@ def _prepare_stock_move_vals(self, first_line, order_lines):
             'company_id': self.company_id.id,
         }
 
+    @api.model
+    def _create_move_lines_for_pos_order(self, moves, set_quantity_done_on_move=False):
+        moves._action_assign()
+        for move_line in moves.move_line_ids:
+            move_line.qty_done = move_line.product_uom_qty
+        mls_vals = []
+        moves_to_set = set()
+        for move in moves:
+            if float_compare(move.product_uom_qty, move.quantity_done, precision_rounding=move.product_uom.rounding) > 0:
+                remaining_qty = move.product_uom_qty - move.quantity_done
+                mls_vals.append(dict(move._prepare_move_line_vals(), qty_done=remaining_qty))
+                moves_to_set.add(move.id)
+        self.env['stock.move.line'].create(mls_vals)
+        if set_quantity_done_on_move:
+            for move in self.env['stock.move'].browse(moves_to_set):
+                move.quantity_done = move.product_uom_qty
+
+    def _create_production_lots_for_pos_order(self, lines):
+        ''' Search for existing lots and create missing ones.
+
+            :param lines: pos order lines with pack lot ids.
+            :type lines: pos.order.line recordset.
+
+            :return stock.production.lot recordset.
+        '''
+        self.ensure_one()
+        existing_lots = self.env['stock.production.lot']
+        if self.picking_type_id.use_existing_lots:
+            lots = lines.pack_lot_ids.filtered(lambda l: l.lot_name)
+            lots_data = set(lots.mapped(lambda l: (l.product_id.id, l.lot_name)))
+            existing_lots = self.env['stock.production.lot'].search([
+                ('company_id', '=', self.company_id.id),
+                ('product_id', 'in', lines.product_id.ids),
+                ('name', 'in', lots.mapped('lot_name')),
+            ])
+            # The previous search may return (product_id.id, lot_name) combinations that have no matching in lines.pack_lot_ids.
+            existing_lots = existing_lots.filtered(lambda l: (l.product_id.id, l.name) in lots_data)
+            if self.picking_type_id.use_create_lots:
+                existing_product_ids = set(existing_lots.product_id.ids)
+                existing_lot_names = set(existing_lots.mapped('name'))
+                missing_lot_values = []
+                for lot in lots:
+                    if lot.lot_name not in existing_lot_names or lot.product_id.id not in existing_product_ids:
+                        missing_lot_values.append({'company_id': self.company_id.id, 'product_id': lot.product_id.id, 'name': lot.lot_name})
+                existing_lots |= self.env['stock.production.lot'].create(missing_lot_values)
+        return existing_lots
```

Comment:
I think it is easier and more clear to use the existing lots_data to know what lots we need to create

```suggestion
        for lot in existing_lot:
            if (lot.product_id.id,l.name) in lots_data:
                valid_lots |= lot
                lots_data.remove((lot.product_id.id,l.name))
            
        if self.picking_type_id.use_create_lots:
            missing_lot_values = []
            for lot in lots_data:
                missing_lot_values.append({'company_id': self.company_id.id, 'product_id': lot[0], 'name': lot[1]})
                valid_lots |= self.env['stock.production.lot'].create(missing_lot_values)
    return valid_lots
```